### PR TITLE
docs(cn): translate content/blog/2022-03-08-react-18-upgrade-guide.md

### PR DESCRIPTION
**Pull request by jsbugwang**
How to Upgrade to React 18 翻译完成。

### CODE REVIEW

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 1]
```diff
@@ -136,51 +136,51 @@ interface MyButtonProps {
 }
 ```
 
-See the [React 18 typings pull request](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210) for a full list of type-only changes. It links to example fixes in library types so you can see how to adjust your code. You can use the [automated migration script](https://github.com/eps1lon/types-react-codemod) to help port your application code to the new and safer typings faster.
+查看 [React 18 typings pull request](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210) 获得只含类型更改（type-only changes）的完整清单。它链到库类型的修复例子（It links to example fixes in library types），你就知道如何调整自己的代码。 你可以使用 [automated migration script](https://github.com/eps1lon/types-react-codemod) 来加速帮你把应用代码迁移到更安全的新类型。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:139]:
```suggestion
查看 [React 18 typings pull request](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210) 获得只含类型更改的完整清单。它链到库类型的修复例子，你就知道如何调整自己的代码。 你可以使用 [自动迁移脚本](https://github.com/eps1lon/types-react-codemod) 来加速帮你把应用代码迁移到更安全的新类型。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 2]
```diff
@@ -136,51 +136,51 @@ interface MyButtonProps {
...
 
-If you find a bug in the typings, please [file an issue](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/new?category=issues-with-a-types-package) in the DefinitelyTyped repo.
+如果你发现了类型的 bug ，请在 DefinitelyTyped repo [提交一个 issue](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/new?category=issues-with-a-types-package)。
 
-## Automatic Batching {#automatic-batching}
+## 自动化批处理（Automatic Batching） {#automatic-batching}
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:143]:
```suggestion
## 自动化批处理 {#automatic-batching}
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 3]
```diff
@@ -136,51 +136,51 @@ interface MyButtonProps {
...
 
-React 18 adds out-of-the-box performance improvements by doing more batching by default. Batching is when React groups multiple state updates into a single re-render for better performance. Before React 18, we only batched updates inside React event handlers. Updates inside of promises, setTimeout, native event handlers, or any other event were not batched in React by default:
+React 18 通过更多默认批处理提供了开箱即用的性能优化。批处理是指 React 把多个状态更新组合为一次重渲染（re-render）从而获得更好性能。React 18 之前，我们仅把 React event handler 的更新做批处理。Promise、setTimeout、native event handler，或者其他事件中的更新，React 默认不进行批处理：
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:145]:
```suggestion
React 18 通过更多默认批处理提供了开箱即用的性能优化。批处理是指 React 把多个状态更新组合为一次重渲染从而获得更好性能。React 18 之前，我们仅把 React event handler 的更新做批处理。Promise、setTimeout、native event handler，或者其他事件中的更新，React 默认不进行批处理：
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 4]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
   flushSync(() => {
     setFlag(f => !f);
   });
-  // React has updated the DOM by now
+  // 现在 React 已更新 DOM
 }
 ```
 
-For more information, see the [Automatic batching deep dive](https://github.com/reactwg/react-18/discussions/21).
+阅读 [深入自动批处理](https://github.com/reactwg/react-18/discussions/21) 获取更多信息。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:200]:
```suggestion
阅读[深入自动批处理](https://github.com/reactwg/react-18/discussions/21) 获取更多信息。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 5]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-## New APIs for Libraries {#new-apis-for-libraries}
+## 为库提供的新 API （New APIs for Libraries） {#new-apis-for-libraries}
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:202]:
```suggestion
## 为库提供的新 API  {#new-apis-for-libraries}
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 6]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-In the React 18 Working Group we worked with library maintainers to create new APIs needed to support concurrent rendering for use cases specific to their use case in areas like styles, and external stores. To support React 18, some libraries may need to switch to one of the following APIs:
+React 18 工作组和库维护者合作，为他们的特定 use case ，例如样式和外部 store（like styles, and external stores），创建了支持并发渲染的新 API。为了支持 React 18，有些库可能需要切换到以下 API 中的一个：
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:204]:
```suggestion
React 18 工作组和库维护者合作，为他们的特定 use case ，例如样式和外部 store ，创建了支持并发渲染的新 API 。为了支持 React 18，有些库可能需要切换到以下 API 中的一个：
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 7]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
-* `useInsertionEffect` is a new hook that allows CSS-in-JS libraries to address performance issues of injecting styles in render. Unless you've already built a CSS-in-JS library we don't expect you to ever use this. This hook will run after the DOM is mutated, but before layout effects read the new layout. This solves an issue that already exists in React 17 and below, but is even more important in React 18 because React yields to the browser during concurrent rendering, giving it a chance to recalculate layout. For more information, see the [Library Upgrade Guide for `<style>`](https://github.com/reactwg/react-18/discussions/110).
+* `useSyncExternalStore` 是一个新的 hook ，通过强制把 store 的更新设置为同步，允许外部 store 支持并发读（concurrent reads）。推荐所有集成了 React 外部状态的库使用这个新 API 。 更多信息请阅读 [useSyncExternalStore overview post](https://github.com/reactwg/react-18/discussions/70) 和 [useSyncExternalStore API details](https://github.com/reactwg/react-18/discussions/86)。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:206]:
```suggestion
* `useSyncExternalStore` 是一个新的 hook ，通过强制把 store 的更新设置为同步，允许外部 store 支持并发读。推荐所有集成了 React 外部状态的库使用这个新 API 。更多信息请阅读 [useSyncExternalStore 综述帖子](https://github.com/reactwg/react-18/discussions/70) 和 [useSyncExternalStore API 细节](https://github.com/reactwg/react-18/discussions/86)。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 8]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* `useInsertionEffect` 是一个新的 hook ，允许 CSS-in-JS 库定位渲染时注入的样式的性能问题。除非你已构建了 CSS-in-JS 库，否则不要用这个 hook 。这个 hook 会在 DOM 可交互后运行，但是在布局 effect 读到新的布局前。这解决了 React 17 及以下版本已存在的一个问题，不过对于 React 18 更重要，因为 React 在并发渲染期间会把控制权交给浏览器，让浏览器有机会重新计算布局。阅读 [Library Upgrade Guide for `<style>`](https://github.com/reactwg/react-18/discussions/110) 获取更多信息。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:207]:
```suggestion
* `useInsertionEffect` 是一个新的 hook ，允许 CSS-in-JS 库定位渲染时注入的样式的性能问题。除非你已构建了 CSS-in-JS 库，否则不要用这个 hook 。这个 hook 会在 DOM 可交互后运行，但是在布局 effect 读到新的布局前。这解决了 React 17 及以下版本已存在的一个问题，不过对于 React 18 更重要，因为 React 在并发渲染期间会把控制权交给浏览器，让浏览器有机会重新计算布局。阅读 [`<style>` 的库升级指南](https://github.com/reactwg/react-18/discussions/110) 获取更多信息。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 9]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-React 18 also introduces new APIs for concurrent rendering such as `startTransition`, `useDeferredValue` and `useId`, which we share more about in the [release post](/blog/2022/03/29/react-v18.html).
+React 18 还引入了 `startTransition`、`useDeferredValue` 和 `useId` 等并发渲染的新 API，在 [release post](/blog/2022/03/29/react-v18.html) 中我们有更详细介绍。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:209]:
```suggestion
React 18 还引入了 `startTransition`、`useDeferredValue` 和 `useId` 等并发渲染的新 API，在 [发布公告](/blog/2022/03/29/react-v18.html) 中我们有更详细介绍。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 10]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+未来我们会添加一个特性，允许 React 增加和删除 UI 部分（add and remove sections of the UI），同时保留状态。例如，用户切换标签走了再回来，React 应该能立即展示之前的屏幕。为了实现这点，React 需用之前的组件状态卸载和重挂载组件树。
 
-This feature will give React better performance out-of-the-box, but requires components to be resilient to effects being mounted and destroyed multiple times. Most effects will work without any changes, but some effects assume they are only mounted or destroyed once.
+这个特性给 React 带来了开箱即用的更佳性能，但是需要组件适应 effect 多次挂载和销毁。大部分 effect 不需要任何修改就能工作，有些则假设他们只会挂载和销毁一次。
 
-To help surface these issues, React 18 introduces a new development-only check to Strict Mode. This new check will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount.
+为了让这些问题暴露，React 18 给严格模式引入了新的仅开发模式检查（development-only check）。这项新检查会自动对每个组件卸载和重挂载，第二次 mount 的时候恢复之前的状态。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:217]:
```suggestion
为了让这些问题暴露，React 18 给严格模式引入了新的仅开发模式检查。这项新检查会自动对每个组件卸载和重挂载，第二次挂载的时候恢复之前的状态。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 11]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-Before this change, React would mount the component and create the effects:
+这次更改前，React 按如下方式 mount 组件和创建 effect ：
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:219]:
```suggestion
这次更改前，React 按如下方式挂载组件和创建 effect ：
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 12]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
 ```
-* React mounts the component.
-    * Layout effects are created.
-    * Effect effects are created.
+* React mount 组件。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:222]:
```suggestion
* React 挂载组件。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 13]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
-* React mounts the component.
-    * Layout effects are created.
-    * Effect effects are created.
-* React simulates unmounting the component.
-    * Layout effects are destroyed.
-    * Effects are destroyed.
-* React simulates mounting the component with the previous state.
-    * Layout effect setup code runs
-    * Effect setup code runs
+* React mounts 组件。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:230]:
```suggestion
* React 挂载组件。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 14]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+    * Layout effect 创建。
+    * Effect effect 创建。
+* React 模拟卸载组件。
+    * Layout effect 销毁。
+    * Effects 销毁。
+* React 模拟用之前的状态 mount 组件。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:236]:
```suggestion
* React 模拟用之前的状态挂载组件。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 15]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+    * 运行 Layout effect 的 setup 代码
+    * 运行 Effect 的 setup 代码
 ```
 
-For more information, see the Working Group posts for [Adding Reusable State to StrictMode](https://github.com/reactwg/react-18/discussions/19) and [How to support Reusable State in Effects](https://github.com/reactwg/react-18/discussions/18).
+请阅读工作组关于 [Adding Reusable State to StrictMode](https://github.com/reactwg/react-18/discussions/19) 和 [How to support Reusable State in Effects](https://github.com/reactwg/react-18/discussions/18) 的文章获取更多信息。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:241]:
```suggestion
请阅读工作组关于 [在严格模式下添加可重用状态](https://github.com/reactwg/react-18/discussions/19) 和 [如何在 Effect 中支持可重用的状态](https://github.com/reactwg/react-18/discussions/18) 的文章获取更多信息。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 16]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-## Configuring Your Testing Environment {#configuring-your-testing-environment}
+## 配置你的测试环境 {#configuring-your-testing-environment}
 
-When you first update your tests to use `createRoot`, you may see this warning in your test console:
+当你第一次更新你的测试用例为使用 `createRoot` ，你可能在测试控制台看到这样的警告：
 
-> The current testing environment is not configured to support act(...)
+> 当前测试环境未配置为支持 act(...)
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:247]:
```suggestion
> The current testing environment is not configured to support act(...)
```

Comment (daochouwangu) [reply content/blog/2022-03-08-react-18-upgrade-guide.md:247]: 这里是控制台错误，是不是不翻译比较好

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 17]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+在运行测试前设置 `globalThis.IS_REACT_ACT_ENVIRONMENT` 为 `true` 可以修复这问题：
 
 ```js
-// In your test setup file
+// 你的测试 setup 文件
 globalThis.IS_REACT_ACT_ENVIRONMENT = true;
 ```
 
-The purpose of the flag is to tell React that it's running in a unit test-like environment. React will log helpful warnings if you forget to wrap an update with `act`.
+此标识的目的是告诉 React 运行在类似单元测试的环境下（unit test-like environment）。如果你忘记用 `act` 包裹更新，React 会记录有用的警告。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:256]:
```suggestion
此标识的目的是告诉 React 运行在类似单元测试的环境下。如果你忘记用 `act` 包裹更新，React 会记录有用的警告。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 18]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-You can also set the flag to `false` to tell React that `act` isn't needed. This can be useful for end-to-end tests that simulate a full browser environment.
+你也可以把标识置为 `false` 告诉 React 不需要用 `act` 。 这对于模拟完整浏览器环境的 e2e 测试很有用。
 
-Eventually, we expect testing libraries will configure this for you automatically. For example, the [next version of React Testing Library has built-in support for React 18](https://github.com/testing-library/react-testing-library/issues/509#issuecomment-917989936) without any additional configuration.
+最后，我们期望测试库帮你自动配置。例如 [next version of React Testing Library has built-in support for React 18](https://github.com/testing-library/react-testing-library/issues/509#issuecomment-917989936) 不需要任何额外配置。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:260]:
```suggestion
最后，我们期望测试库帮你自动配置。例如 [下一版本的 React 测试库已内置对 React 18 的支持](https://github.com/testing-library/react-testing-library/issues/509#issuecomment-917989936) 不需要任何额外配置。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 19]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-[More background on the `act` testing API and related changes](https://github.com/reactwg/react-18/discussions/102) is available in the working group.
+这里还有工作组[关于 `act` 测试 API 和相关更改的更多背景](https://github.com/reactwg/react-18/discussions/102)。
 
-## Dropping Support for Internet Explorer {#dropping-support-for-internet-explorer}
+## 不再支持 Internet Explorer {#dropping-support-for-internet-explorer}
 
-In this release, React is dropping support for Internet Explorer, which is [going out of support on June 15, 2022](https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge). We’re making this change now because new features introduced in React 18 are built using modern browser features such as microtasks which cannot be adequately polyfilled in IE.
+从这个版本起，React 放弃了对 IE 的支持，IE[从2022-06-15不再支持](https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge)了。我们现在做出这个改变，是因为 React 18 引入的新特性是基于现代浏览器的新特性，比如微任务，而这些在 IE 不能很好的用 polyfill 支持。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:266]:
```suggestion
从这个版本起，React 放弃了对 IE 的支持，IE [从2022-06-15不再支持](https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge)了。我们现在做出这个改变，是因为 React 18 引入的新特性是基于现代浏览器的新特性，比如微任务，而这些在 IE 不能很好的用 polyfill 支持。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 20]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
 
-If you need to support Internet Explorer we recommend you stay with React 17.
+如果你需要支持 IE ，我们建议你停留在 React 17 。
 
-## Deprecations {#deprecations}
+## 废弃（Deprecations） {#deprecations}
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:270]:
```suggestion
## 废弃 {#deprecations}
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 21]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
-* `react-dom`: `ReactDOM.renderSubtreeIntoContainer` has been deprecated.
-* `react-dom/server`: `ReactDOMServer.renderToNodeStream` has been deprecated.
+* `react-dom`: `ReactDOM.render` 已废弃。继续使用会发出警告，并且应用运行在 React 17 模式。
+* `react-dom`: `ReactDOM.hydrate` 已废弃。继续使用会发出警告，并且应用运行在 React 17 模式。
+* `react-dom`: `ReactDOM.unmountComponentAtNode` 已废弃。
+* `react-dom`: `ReactDOM.renderSubtreeIntoContainer` 已废弃。
+* `react-dom/server`: `ReactDOMServer.renderToNodeStream` 已废弃。
 
-## Other Breaking Changes {#other-breaking-changes}
+## 其他破坏性更改（Other Breaking Changes） {#other-breaking-changes}
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:278]:
```suggestion
## 其他破坏性更改 {#other-breaking-changes}
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 22]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
-* **Layout Effects with Suspense**: When a tree re-suspends and reverts to a fallback, React will now clean up layout effects, and then re-create them when the content inside the boundary is shown again. This fixes an issue which prevented component libraries from correctly measuring layout when used with Suspense.
-* **New JS Environment Requirements**: React now depends on modern browsers features including `Promise`, `Symbol`, and `Object.assign`. If you support older browsers and devices such as Internet Explorer which do not provide modern browser features natively or have non-compliant implementations, consider including a global polyfill in your bundled application.
+* **一致的 useEffect 时序（Consistent useEffect timing）**: 如果更新是由独立的用户输入事件，如 click 或 keydown 事件，则 React 现在总是同步刷新 （flush）effect 函数。而以前行为并非总是可预测的和一致的。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:280]:
```suggestion
* **一致的 useEffect 时序**: 如果更新是由独立的用户输入事件，如 click 或 keydown 事件，则 React 现在总是同步刷新 effect 函数。而以前行为并非总是可预测的和一致的。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 23]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **更严格的 hydration 错误（Stricter hydration errors）**: 由于缺失的或额外的文本内容现在被当做错误而非警告，因此 Hydration 匹配不上了。React 不再尝试在客户端插入或删除节点对单独节点进行 "patch up" ，以匹配服务端的内容，而是回退到客户端，渲染组件树最近的 `<Suspense>` 边界。这样保证 hydrate 后树的一致性，避免潜在的由 hydration 不匹配导致的隐私和安全漏洞 。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:281]:
```suggestion
* **更严格的 hydration 错误**: 由于缺失的或额外的文本内容现在被当做错误而非警告，因此 hydration 匹配不上了。React 不再尝试在客户端插入或删除节点对单独节点进行 "patch up" ，以匹配服务端的内容，而是回退到客户端，渲染组件树最近的 `<Suspense>` 边界。这样保证 hydrate 后树的一致性，避免潜在的由 hydration 不匹配导致的隐私和安全漏洞 。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 24]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **Suspense 树总是一致的：** 如果一个组件在完全添加到组件树前挂起了，React 会以不完整状态将其添加到组件树，或者触发其副作用。React 会完全抛弃新的树，等待异步操作完成，然后尝试从头开始渲染（retry rendering again from scratch）。React 用并发渲染进行重试，并且不阻塞浏览器。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:282]:
```suggestion
* **Suspense 树总是一致的：** 如果一个组件在完全添加到组件树前挂起了，React 会以不完整状态将其添加到组件树，或者触发其副作用。React 会完全抛弃新的树，等待异步操作完成，然后尝试从头开始渲染。React 用并发渲染进行重试，并且不阻塞浏览器。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 25]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **Layout Effects with Suspense**: 当组件树重新挂起（re-suspend）并回退到兜底时（reverts to a fallback），React 会清理布局 effect，然后当边界内的内容重新显示时，重新创建它们。这解决了一个问题，这个问题会阻止组件库在使用 Suspense 时正确测量布局（measuring layout）。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:283]:
```suggestion
* **Layout Effects with Suspense** : 当组件树重新挂起并回退到兜底时，React 会清理布局 effect ，然后当边界内的内容重新显示时，重新创建它们。这解决了一个问题，这个问题会阻止组件库在使用 Suspense 时正确测量布局。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 26]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **新的 JS 环境要求**: React 现在依赖现代浏览器的特性，包括 `Promise`、`Symbol` 和 `Object.assign`。如果你支持老的浏览器和设备，例如 IE ，对现代浏览器特性缺乏原生支持或者有兼容性问题，可以考虑在应用 bundle 增加全局 polyfill 。
 
-## Other Notable Changes {#other-notable-changes}
+## 其他重要更改（Other Notable Changes） {#other-notable-changes}
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:286]:
```suggestion
## 其他重要更改 {#other-notable-changes}
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 27]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
-* **No suppression of console logs:** When you use Strict Mode, React renders each component twice to help you find unexpected side effects. In React 17, we've suppressed console logs for one of the two renders to make the logs easier to read. In response to [community feedback](https://github.com/facebook/react/issues/21783) about this being confusing, we've removed the suppression. Instead, if you have React DevTools installed, the second log's renders will be displayed in grey, and there will be an option (off by default) to suppress them completely.
-* **Improved memory usage:** React now cleans up more internal fields on unmount, making the impact from unfixed memory leaks that may exist in your application code less severe.
+* **组件现在可以渲染 `undefined`:** 如果组件返回 `undefined`，React 不再警告。这样组件的返回值与组件树中间允许的值（values that are allowed in the middle of a component tree）保持一致。我们建议使用 linter 来避免在 JSX 之前忘记 `return` 语句的错误。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:290]:
```suggestion
* **组件现在可以渲染 `undefined`:** 如果组件返回 `undefined`，React 不再警告。这样组件的返回值与组件树中间允许的值保持一致。我们建议使用 linter 来避免在 JSX 之前忘记 `return` 语句的错误。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 28]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **在测试中, `act` 警告是可选的:** 如果你在运行 e2e 测试，`act` 警告是不必要的。我们引入了一个 [opt-in](https://github.com/reactwg/react-18/discussions/102) 机制，这样你可以在单元测试时启用，对于单元测试是有用和有收益的。
+* **未挂载组件 `setState` 不再有警告:** 以前, 当你在未挂载组件调用 `setState` 时，React 警告内存泄露。该警告是为了订阅添加的，然后人们遇到这种情况时，大部分时候都是正常情况，而解决方法（workaround）让代码更糟糕。 我们已[删除](https://github.com/facebook/react/pull/22114)这个警告。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:292]:
```suggestion
* **未挂载组件 `setState` 不再有警告:** 以前, 当你在未挂载组件调用 `setState` 时，React 警告内存泄露。该警告是为了订阅添加的，然后人们遇到这种情况时，大部分时候都是正常情况，而解决方法让代码更糟糕。 我们已[删除](https://github.com/facebook/react/pull/22114)这个警告。
```

File: content/blog/2022-03-08-react-18-upgrade-guide.md [hunk 29]
```diff
@@ -189,115 +189,115 @@ function handleClick() {
...
+* **不抑制控制台日志（No suppression of console logs）：** 严格模式下, React 每个组件渲染2次，帮你发现意外的副作用。在 React 17 中我们抑制（suppress）了其中一次渲染的控制台日志，让日志更易读。根据[社区反馈](https://github.com/facebook/react/issues/21783)，这个功能让人困惑，我们已删除了这个功能。现在如果你安装了 React DevTools，第二条日志显示成灰色的，同时提供一个选项（默认不开启）彻底抑制日志。
```
Comment (daochouwangu) [inline content/blog/2022-03-08-react-18-upgrade-guide.md:293]:
```suggestion
* **不抑制控制台日志：** 严格模式下, React 每个组件渲染2次，帮你发现意外的副作用。在 React 17 中我们抑制了其中一次渲染的控制台日志，让日志更易读。根据[社区反馈](https://github.com/facebook/react/issues/21783)，这个功能让人困惑，我们已删除了这个功能。现在如果你安装了 React DevTools，第二条日志显示成灰色的，同时提供一个选项（默认不开启）彻底抑制日志。 
```

**Review by daochouwangu — CHANGES_REQUESTED** [review]

**Comment by awxiaoxian2020** [comment]
@jsbugwang Could you please handle with reviews?

**Comment by jsbugwang** [comment]
> 

suggestion resolved

**Review by daochouwangu — CHANGES_REQUESTED** [review]

**Review by daochouwangu — APPROVED** [review]

**Review by awxiaoxian2020 — APPROVED** [review]

**Pull request merged**